partiql-ast/src/main/java/org/partiql/ast/literal/Literal.java

```diff
@@ -0,0 +1,125 @@
+package org.partiql.ast.literal;
+
+import org.jetbrains.annotations.NotNull;
+import org.partiql.ast.AstNode;
+import org.partiql.ast.AstVisitor;
+import org.partiql.ast.DataType;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.Collection;
+import java.util.Collections;
+
+public abstract class Literal extends AstNode {
```

Comment:
Latest design currently models things a bit differently than before. Key summary is
- fat interface w/ a `LiteralKind` enum. Includes relevant constructors and property accessor methods. much like PType, Datum, AnyElement, and Calcite's [SqlLiteral](https://calcite.apache.org/javadocAggregate/org/apache/calcite/sql/SqlLiteral.html)
- Three numeric variants (approx, int, exact) with each containing just a string
  - The parser will still perform some type ascription
  - Modeling certain numerics with a string proved less cumbersome. E.g.
    - `123` vs `123.` -- BigDecimal does not actually preserve whether there was a period or not
    - `123e0` vs `123.e0` -- BigDecimal for the mantissa of an approximate does not preserve whether there was a period or not
    - `123e0` vs `123.e+0` -- a long/integer for the exponent does not preserve whether there was an explicit `+`
  - Simplifying to just a string allows us to hold off on any eager computation
- Date/time/timestamp represented as a typed string
  - The parser already performs some validations and basic type ascription. 
  - One potential issue with eagerly converting the fields into java is all of the additional properties that would need added. E.g. date would need year, month, day. Time would need hour, minute, second, precision, timezone. Timestamp would need all of date and time's fields. We would need to at some point convert these AST fields to the plan's representation.

Reply:
> One potential issue with eagerly converting the fields into java is all of the additional properties that would need added. E.g. date would need year, month, day. Time would need hour, minute, second, precision, timezone. Timestamp would need all of date and time's fields. We would need to at some point convert these AST fields to the plan's representation.


That's a good point, and I didn't think of that. We wouldn't be able to zero-pad and would need to insert nulls for those fields. Seems like a lot of extra work (on read aka scribe) when a (rule,string) is sufficient for capturing syntax which ultimately is the goal.

Reply:
> capturing syntax which ultimately is the goal

Yeah that was my mindset w/ this latest refactor -- AST at the end of the day just captures syntax with some slight validation w/ the parser. I think most users are not actually performing computation directly on the AST. Computation (like in our case) will come later in the plan or even pushed all the way to the evaluator.